exonum/src/api/error.rs

```diff
@@ -16,10 +16,118 @@
 
 use failure::Fail;
 use serde::Serialize;
+
+use std::collections::HashMap;
 use std::io;
 
 use crate::node::SendError;
 
+/// List of possible HTTP response codes.
+#[derive(Debug, Serialize)]
+pub enum HttpCode {
+    #[allow(missing_docs)]
+    BadRequest = 400,
+    #[allow(missing_docs)]
+    NotFound = 404,
+    #[allow(missing_docs)]
+    NotImplemented = 501,
+}
+
+/// API HTTP error struct.
+#[derive(Fail, Debug, Serialize)]
+pub struct ApiError {
+    /// HTTP error code.
+    #[serde(skip)]
+    pub http_code: HttpCode,
+    /// A URI reference to the documentation or possible solutions for the problem.
+    #[serde(rename = "type")]
+    pub error_type: String,
+    /// Short description of the error.
+    pub title: String,
+    /// Detailed description of the error.
+    pub detail: String,
+    /// HashMap with additional parameters.
+    pub params: HashMap<String, String>,
+    /// Source of the error.
+    pub source: String,
+    /// Internal error code.
+    pub error_code: i8,
+}
+
+impl std::fmt::Display for ApiError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&serde_json::to_string(self).unwrap())
+    }
+}
+
+impl ApiError {
+    fn default() -> Self {
+        Self {
+            http_code: HttpCode::NotImplemented,
+            error_type: String::new(),
+            title: String::new(),
+            detail: String::new(),
+            params: HashMap::new(),
+            source: String::new(),
+            error_code: 0,
+        }
+    }
+
+    /// Builds a BadRequest error.
+    #[allow(non_snake_case)]
+    pub fn BadRequest() -> Self {
+        Self {
+            http_code: HttpCode::BadRequest,
+            ..Self::default()
+        }
+    }
+
+    /// Builds a NotFound error.
+    #[allow(non_snake_case)]
+    pub fn NotFound() -> Self {
+        Self {
+            http_code: HttpCode::NotFound,
+            ..Self::default()
+        }
+    }
+
+    /// Sets `error_type` of an error.
+    pub fn error_type(mut self, error_type: String) -> Self {
```

Comment:
Maybe `&str` would be better from the UX point of view?
```rust
.error_type("Some type of error")
```
vs
```rust
.error_type("Some type of error".to_string()) 
``` 
What do you think ?

Reply:
I already suggested to use `Into<String>`: https://github.com/exonum/exonum/pull/1696#discussion_r365887349

Reply:
With Into<String> will be almost the same.

Reply:
@popzxc, even better 👍🏻

Reply:
@Kormick `Into` allow you to use `&str` and `String` at the same time. Take a look: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=236b56b94c702cea1f93ee64e4443525